gobblin-core/src/main/java/gobblin/source/extractor/extract/kafka/KafkaSource.java

```diff
@@ -83,6 +95,211 @@
     }
   }
 
+  /**
+   * Group workUnits into multiWorkUnits. Each workUnit corresponds to a (topic, partition).
+   * The goal is to group the workUnits as evenly as possible into numOfMultiWorkunits multiWorkunits,
+   * while preferring to put partitions of the same topic into the same multiWorkUnits (in order to
+   * avoid generating many small mapper output files).
+   *
+   * The algorithm is to first group workUnits into approximately 3 * numOfMultiWorkunits groups using
+   * best-fit-decreasing, such that partitions of the same topic are in the same group. Then, 
+   * assemble these groups into multiWorkunits using worst-fit-decreasing.
+   *
+   * The reason behind 3 is that the worst-fit-decreasing algorithm should work well if the number of
+   * items is more than 3 times the number of bins.
+   *
+   * The input workUnits are partitioned by topic.
+   */
+  private List<WorkUnit> getMultiWorkunits(List<List<WorkUnit>> workUnits, int numOfMultiWorkunits) {
+    Preconditions.checkArgument(numOfMultiWorkunits >= 1);
+
+    List<List<WorkUnitWithEstSize>> workUnitsWithEstSize = getWorkUnitsWithEstSize(workUnits);
+    long totalEstimatedDataSize = calcTotalEstSize(workUnitsWithEstSize);
+    long avgGroupSize = (long) ((double) totalEstimatedDataSize / (double) numOfMultiWorkunits / 3.0);
+
+    List<WorkUnitGroup> groups = Lists.newArrayList();
+    for (List<WorkUnitWithEstSize> workUnitsWithEstSizeForTopic : workUnitsWithEstSize) {
+      long estimatedDataSizeForTopic = calcTotalEstSizeForTopic(workUnitsWithEstSizeForTopic);
+      if (estimatedDataSizeForTopic < avgGroupSize) {
+
+        // If the total estimated size of a topic is smaller than group size, put all partitions of this
+        // topic in a single group.
+        groups.add(new WorkUnitGroup(workUnitsWithEstSizeForTopic));
+      } else {
+
+        // Use best-fit-decreasing to group workunits for a topic into multiple groups.
+        groups.addAll(bestFitDecreasingBinPacking(workUnitsWithEstSizeForTopic, avgGroupSize));
+      }
+    }
+
+    return worstFitDecreasingBinPacking(groups, numOfMultiWorkunits);
+  }
+
+  /**
+   * Pack a list of WorkUnitGroups into multiWorkUnits, using the worst-fit-decreasing algorithm,
+   * i.e., for each group, assign it to the multiWorkUnit with the lightest load.
+   */
+  private static List<WorkUnit> worstFitDecreasingBinPacking(List<WorkUnitGroup> groups, int numOfMultiWorkUnits) {
+    List<MultiWorkUnit> multiWorkunits = Lists.newArrayListWithCapacity(numOfMultiWorkUnits);
+    List<Long> loads = Lists.newArrayListWithCapacity(numOfMultiWorkUnits);
```

Comment:
Again, you can set the load as a property in the `MultiWorkUnit` so you don't need to maintain a separate list.